# Computational Theory Assessment-G00413222

## Introduction

This notebook contains my work for the Computational Theory assessment. The main focus of the assessment is the SHA-256 hashing algorithm, following the NIST Secure Hash Standard (FIPS 180-4).

The work is divided into several problems. I will first look at how SHA-256 data can be represented in Python before implementing the required bitwise operations, constants, message padding, compression function, and finally the complete SHA-256 algorithm.

Each section includes explanations and Python code, along with tests where appropriate.

## Project Structure

- **Problem 1**: Representing SHA-256 data, including 32-bit words, input messages, 512-bit blocks, the 256-bit hash, and big-endian integers.
- **Problem 2**: Implementing the SHA-256 bitwise functions using NumPy.
- **Problem 3**: Generating the SHA-256 constants from prime numbers and their cube roots.
- **Problem 4**: Padding and parsing input messages into 512-bit blocks.
- **Problem 5**: Implementing the SHA-256 compression function.
- **Problem 6**: Combining the previous components to implement and test the complete SHA-256 algorithm.

## Libraries

This assessment uses NumPy for working with 32-bit unsigned integers and arrays. These are useful for representing the words and values used by SHA-256.

In [2]:
# Import NumPy
import numpy as np

## Problem 1: Representing SHA-256 Data

SHA-256 works with fixed-size pieces of data, so it is important to represent these values correctly before implementing the algorithm. This problem looks at 32-bit words, sequences of 32-bit words, input messages, 512-bit blocks, the 256-bit hash value, and big-endian integers.


### 32-bit Words

SHA-256 processes data using 32-bit words. A 32-bit unsigned integer can represent values from \(0\) to \(2^{32}-1\). NumPy's `uint32` type can be used to represent these values explicitly.

Using a fixed-width 32-bit type is important because the SHA-256 algorithm defines its operations using 32-bit words.


In [3]:
# Create a 32-bit unsigned integer
word = np.uint32(42)

print("Value:", word)
print("Type:", type(word))
print("Maximum value:", np.iinfo(np.uint32).max)

Value: 42
Type: <class 'numpy.uint32'>
Maximum value: 4294967295


In [7]:
# Test the maximum value of a 32-bit unsigned integer
maximum_word = np.uint32(2**32 - 1)

print("Maximum 32-bit value:", maximum_word)
print("Hexadecimal:", hex(int(maximum_word)))

Maximum 32-bit value: 4294967295
Hexadecimal: 0xffffffff


The test shows that a 32-bit unsigned integer can store values up to \(2^{32}-1\), which is `4294967295`. The hexadecimal representation `0xffffffff` shows all 32 bits set to 1. This confirms the range of the `uint32` representation being used for SHA-256 words.
